File: rust/src/leet_code/678.rs
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let (mut lo, mut hi) = (0, 0);
        for c in s.chars() {
            match c {
                '(' => {
                    lo += 1;
                    hi += 1;
                }
                ')' => {
                    if lo > 0 {
                        lo -= 1;
                    }
                    hi -= 1;
                }
                _ => {
                    if lo > 0 {
                        lo -= 1;
                    }
                    hi += 1;
                }
            }
            if hi < 0 {
                return false;
            }
        }
        lo == 0
    }
}

File: python/leet_code/678.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        if not s:
            return True
        stack = []
        stars = []
        for i, c in enumerate(s):
            if c == '(':
                stack.append(i)
            elif c == '*':
                stars.append(i)
            else:
                if stack:
                    stack.pop()
                elif stars:
                    stars.pop()
                else:
                    return False
        while stack and stars:
            if stack[-1] > stars[-1]:
                return False
            stack.pop()
            stars.pop()
        return not stack
